src/storage/src/sink/kafka.rs

```diff
@@ -1218,12 +1274,15 @@ fn encode_collection<G: Scope>(
                         value_schema_id,
                     ))
                 }
-                KafkaSinkFormat::Json => Box::new(JsonEncoder::new(
-                    key_desc,
-                    value_desc,
-                    matches!(envelope, SinkEnvelope::Debezium),
-                )),
             };
+            // If the key encoder is None, it means we either don't have a key or the key format
+            // is avro and should share the value encoder. In either case we can use the value
+            // encoder as the key encoder.
```

Comment:
We should, but I'm matching the semantics of the previous implementation here where we try to avoid any errors during the actual encoding step below due to your comment:
```
            // !IMPORTANT!
            // Correctness of this operator relies on no fallible operations happening after this
            // point. This is a temporary workaround of build_fallible's bad interaction of owned
            // capabilities and errors.
            // TODO(petrosagg): Make the fallible async operator safe
```
since the `.encode_key_unchecked` is infallible

Reply:
updated so that this will panic if we encounter a key without a key-encoder set